### ISSUE TITLE
regcomp: heap-buffer-overflow read utf8::= (perl-5.30.0)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#17016]
Migrated from [rt.perl.org#134133](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=134133) (status was 'pending release')

Searchable as RT134133$

**Comment by p5pRT** [comment]
### From @Etsukata

PoC

\`\`\`

\[eiichi@&#8203;x1 ~\]$ valgrind perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl \-e
'qr/\\p\{utf8&#8203;::=\}/'                            
==12715== Memcheck\, a memory error detector
==12715== Copyright \(C\) 2002\-2017\, and GNU GPL'd\, by Julian Seward et al\.
==12715== Using Valgrind\-3\.15\.0 and LibVEX; rerun with \-h for copyright info
==12715== Command&#8203;: perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl \-e
qr/\\\\p\{utf8&#8203;::=\}/
==12715==
==12715== Invalid read of size 2
==12715==    at 0x488036&#8203;: Perl\_parse\_uniprop\_string \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x48ED99&#8203;: S\_regclass \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x492095&#8203;: S\_regpiece \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x495E54&#8203;: S\_regbranch \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x4963A7&#8203;: S\_reg \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x49AE68&#8203;: Perl\_re\_op\_compile \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x43928A&#8203;: Perl\_pmruntime \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x46FEBD&#8203;: Perl\_yyparse \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x444AAC&#8203;: perl\_parse \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x42023F&#8203;: main \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==  Address 0x51c8026 is 6 bytes inside a block of size 7 alloc'd
==12715==    at 0x483880B&#8203;: malloc \(vg\_replace\_malloc\.c&#8203;:309\)
==12715==    by 0x4A70D2&#8203;: Perl\_safesysmalloc \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x487E98&#8203;: Perl\_parse\_uniprop\_string \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x48ED99&#8203;: S\_regclass \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x492095&#8203;: S\_regpiece \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x495E54&#8203;: S\_regbranch \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x4963A7&#8203;: S\_reg \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x49AE68&#8203;: Perl\_re\_op\_compile \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x43928A&#8203;: Perl\_pmruntime \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x46FEBD&#8203;: Perl\_yyparse \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x444AAC&#8203;: perl\_parse \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==    by 0x42023F&#8203;: main \(in
/home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
==12715==
Illegal user\-defined property name "utf8&#8203;::=" in regex; marked by \<\-\-
HERE in m/\\p\{utf8&#8203;::=\} \<\-\- HERE / at \-e line 1\.
==12715==
==12715== HEAP SUMMARY&#8203;:
==12715==     in use at exit&#8203;: 153\,845 bytes in 675 blocks
==12715==   total heap usage&#8203;: 890 allocs\, 215 frees\, 190\,688 bytes allocated
==12715==
==12715== LEAK SUMMARY&#8203;:
==12715==    definitely lost&#8203;: 0 bytes in 0 blocks
==12715==    indirectly lost&#8203;: 0 bytes in 0 blocks
==12715==      possibly lost&#8203;: 0 bytes in 0 blocks
==12715==    still reachable&#8203;: 153\,845 bytes in 675 blocks
==12715==         suppressed&#8203;: 0 bytes in 0 blocks
==12715== Rerun with \-\-leak\-check=full to see details of leaked memory
==12715==
==12715== For lists of detected and suppressed errors\, rerun with&#8203;: \-s
==12715== ERROR SUMMARY&#8203;: 1 errors from 1 contexts \(suppressed&#8203;: 0 from 0\)

\`\`\`

**Comment by p5pRT** [comment]
### From @khwilliamson

Resending the below as it did not get attached to the ticket\.

On 5/23/19 11&#8203;:52 PM\, Eiichi Tsukata \(via RT\) wrote&#8203;:
> \# New Ticket Created by  Eiichi Tsukata
> \# Please include the string&#8203;:  \[perl \#134133\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=134133 >
> 
> 
> PoC
> 
> \`\`\`
> 
> \[eiichi@&#8203;x1 ~\]$ valgrind perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl \-e
> 'qr/\\p\{utf8&#8203;::=\}/'
> ==12715== Memcheck\, a memory error detector
> ==12715== Copyright \(C\) 2002\-2017\, and GNU GPL'd\, by Julian Seward et al\.
> ==12715== Using Valgrind\-3\.15\.0 and LibVEX; rerun with \-h for copyright info
> ==12715== Command&#8203;: perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl \-e
> qr/\\\\p\{utf8&#8203;::=\}/
> ==12715==
> ==12715== Invalid read of size 2
> ==12715==    at 0x488036&#8203;: Perl\_parse\_uniprop\_string \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x48ED99&#8203;: S\_regclass \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x492095&#8203;: S\_regpiece \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x495E54&#8203;: S\_regbranch \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x4963A7&#8203;: S\_reg \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x49AE68&#8203;: Perl\_re\_op\_compile \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x43928A&#8203;: Perl\_pmruntime \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x46FEBD&#8203;: Perl\_yyparse \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x444AAC&#8203;: perl\_parse \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x42023F&#8203;: main \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==  Address 0x51c8026 is 6 bytes inside a block of size 7 alloc'd
> ==12715==    at 0x483880B&#8203;: malloc \(vg\_replace\_malloc\.c&#8203;:309\)
> ==12715==    by 0x4A70D2&#8203;: Perl\_safesysmalloc \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x487E98&#8203;: Perl\_parse\_uniprop\_string \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x48ED99&#8203;: S\_regclass \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x492095&#8203;: S\_regpiece \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x495E54&#8203;: S\_regbranch \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x4963A7&#8203;: S\_reg \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x49AE68&#8203;: Perl\_re\_op\_compile \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x43928A&#8203;: Perl\_pmruntime \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x46FEBD&#8203;: Perl\_yyparse \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x444AAC&#8203;: perl\_parse \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==    by 0x42023F&#8203;: main \(in
> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
> ==12715==
> Illegal user\-defined property name "utf8&#8203;::=" in regex; marked by \<\-\-
> HERE in m/\\p\{utf8&#8203;::=\} \<\-\- HERE / at \-e line 1\.
> ==12715==
> ==12715== HEAP SUMMARY&#8203;:
> ==12715==     in use at exit&#8203;: 153\,845 bytes in 675 blocks
> ==12715==   total heap usage&#8203;: 890 allocs\, 215 frees\, 190\,688 bytes allocated
> ==12715==
> ==12715== LEAK SUMMARY&#8203;:
> ==12715==    definitely lost&#8203;: 0 bytes in 0 blocks
> ==12715==    indirectly lost&#8203;: 0 bytes in 0 blocks
> ==12715==      possibly lost&#8203;: 0 bytes in 0 blocks
> ==12715==    still reachable&#8203;: 153\,845 bytes in 675 blocks
> ==12715==         suppressed&#8203;: 0 bytes in 0 blocks
> ==12715== Rerun with \-\-leak\-check=full to see details of leaked memory
> ==12715==
> ==12715== For lists of detected and suppressed errors\, rerun with&#8203;: \-s
> ==12715== ERROR SUMMARY&#8203;: 1 errors from 1 contexts \(suppressed&#8203;: 0 from 0\)
> 
> \`\`\`
> 
> 

This is fixed by the attached file

**Comment by p5pRT** [comment]
### From @khwilliamson

<details><summary>0001-PATCH-perl-134133-read-beyond-end-of-buffer.patch</summary>

```diff
>From ee6f83153ea3eca198f39b966057339418840a7f Mon Sep 17 00:00:00 2001
From: Karl Williamson <khw@cpan.org>
Date: Fri, 24 May 2019 09:01:46 -0600
Subject: [PATCH 1/2] PATCH: [perl #134133] read beyond end of buffer

The code was using the wrong limit variable.
---
 regcomp.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/regcomp.c b/regcomp.c
index 9bd6dd3739..51e286bc20 100644
--- a/regcomp.c
+++ b/regcomp.c
@@ -22857,7 +22857,7 @@ Perl_parse_uniprop_string(pTHX_
         /* Certain properties whose values are numeric need special handling.
          * They may optionally be prefixed by 'is'.  Ignore that prefix for the
          * purposes of checking if this is one of those properties */
-        if (memBEGINPs(lookup_name, name_len, "is")) {
+        if (memBEGINPs(lookup_name, j, "is")) {
             lookup_offset = 2;
         }
 
-- 
2.17.1



```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @khwilliamson

On 6/27/19 12&#8203;:26 PM\, Karl Williamson wrote&#8203;:
> Resending the below as it did not get attached to the ticket\.
> 
> On 5/23/19 11&#8203;:52 PM\, Eiichi Tsukata \(via RT\) wrote&#8203;:
>> \# New Ticket Created by  Eiichi Tsukata
>> \# Please include the string&#8203;:  \[perl \#134133\]
>> \# in the subject line of all future correspondence about this issue\.
>> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=134133 >
>>
>>
>> PoC
>>
>> \`\`\`
>>
>> \[eiichi@&#8203;x1 ~\]$ valgrind perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl \-e
>> 'qr/\\p\{utf8&#8203;::=\}/'
>> ==12715== Memcheck\, a memory error detector
>> ==12715== Copyright \(C\) 2002\-2017\, and GNU GPL'd\, by Julian Seward et al\.
>> ==12715== Using Valgrind\-3\.15\.0 and LibVEX; rerun with \-h for 
>> copyright info
>> ==12715== Command&#8203;: perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl \-e
>> qr/\\\\p\{utf8&#8203;::=\}/
>> ==12715==
>> ==12715== Invalid read of size 2
>> ==12715==    at 0x488036&#8203;: Perl\_parse\_uniprop\_string \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x48ED99&#8203;: S\_regclass \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x492095&#8203;: S\_regpiece \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x495E54&#8203;: S\_regbranch \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x4963A7&#8203;: S\_reg \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x49AE68&#8203;: Perl\_re\_op\_compile \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x43928A&#8203;: Perl\_pmruntime \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x46FEBD&#8203;: Perl\_yyparse \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x444AAC&#8203;: perl\_parse \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x42023F&#8203;: main \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==  Address 0x51c8026 is 6 bytes inside a block of size 7 alloc'd
>> ==12715==    at 0x483880B&#8203;: malloc \(vg\_replace\_malloc\.c&#8203;:309\)
>> ==12715==    by 0x4A70D2&#8203;: Perl\_safesysmalloc \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x487E98&#8203;: Perl\_parse\_uniprop\_string \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x48ED99&#8203;: S\_regclass \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x492095&#8203;: S\_regpiece \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x495E54&#8203;: S\_regbranch \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x4963A7&#8203;: S\_reg \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x49AE68&#8203;: Perl\_re\_op\_compile \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x43928A&#8203;: Perl\_pmruntime \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x46FEBD&#8203;: Perl\_yyparse \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x444AAC&#8203;: perl\_parse \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==    by 0x42023F&#8203;: main \(in
>> /home/eiichi/perl5/perlbrew/perls/perl\-5\.30\.0/bin/perl\)
>> ==12715==
>> Illegal user\-defined property name "utf8&#8203;::=" in regex; marked by \<\-\-
>> HERE in m/\\p\{utf8&#8203;::=\} \<\-\- HERE / at \-e line 1\.
>> ==12715==
>> ==12715== HEAP SUMMARY&#8203;:
>> ==12715==     in use at exit&#8203;: 153\,845 bytes in 675 blocks
>> ==12715==   total heap usage&#8203;: 890 allocs\, 215 frees\, 190\,688 bytes 
>> allocated
>> ==12715==
>> ==12715== LEAK SUMMARY&#8203;:
>> ==12715==    definitely lost&#8203;: 0 bytes in 0 blocks
>> ==12715==    indirectly lost&#8203;: 0 bytes in 0 blocks
>> ==12715==      possibly lost&#8203;: 0 bytes in 0 blocks
>> ==12715==    still reachable&#8203;: 153\,845 bytes in 675 blocks
>> ==12715==         suppressed&#8203;: 0 bytes in 0 blocks
>> ==12715== Rerun with \-\-leak\-check=full to see details of leaked memory
>> ==12715==
>> ==12715== For lists of detected and suppressed errors\, rerun with&#8203;: \-s
>> ==12715== ERROR SUMMARY&#8203;: 1 errors from 1 contexts \(suppressed&#8203;: 0 from 0\)
>>
>> \`\`\`
>>
>>
> 
> This is fixed by the attached file
> 
> 

I'm trying to figure out if this is a real potential attack vector\. 
This is malloc'd space at the top of the function\.  It's mallocd to 
name\_len bytes\, but isn't necessarily initialized to the full amount\.

I don't understand how this actually is generating this valgrind error\. 
The underlying macro expands to a memcmp of name\_len bytes\, which 
shouldn't go off the end of the mallocd space\.  But it could certainly 
read data that didn't get initialized\.

On \#irc\, people said this kind of error only comes from reading an 
invalid address\.

**Comment by p5pRT** [comment]
### From @hvds

On Thu\, 27 Jun 2019 12&#8203;:31&#8203;:30 \-0700\, public@&#8203;khwilliamson\.com wrote&#8203;:
> I'm trying to figure out if this is a real potential attack vector\. 
> This is malloc'd space at the top of the function\.  It's mallocd to 
> name\_len bytes\, but isn't necessarily initialized to the full amount\.
> 
> I don't understand how this actually is generating this valgrind error\. 
> The underlying macro expands to a memcmp of name\_len bytes\, which 
> shouldn't go off the end of the mallocd space\.  But it could certainly 
> read data that didn't get initialized\.

I think lookup\_name has been moved since it was malloced\, probably by&#8203;:
&nbsp;  lookup\_name \+=  STRLENs\("utf8&#8203;::"\);

\.\. which would explain valgrind's diagnosis&#8203;:
&nbsp;  Address 0x51c8026 is 6 bytes inside a block of size 7

In any case\, reading \(but not leaking information about\) some bytes of malloc bookkeeping doesn't sound to me like much of an attack vector\. I think worst case is that we get unlucky with the bookkeeping bytes causing the memBEGINPs to return TRUE; I'd expect that to be only a bug though\, not a security issue\.

Briefly looking at other uses of name\_len\, this one might also be dodgy&#8203;:

&nbsp;    /\* If the original input began with 'In' or 'Is'\, it could be a subroutine
&nbsp;     \* call to a user\-defined property instead of a Unicode property name\. \*/
&nbsp;    if \(    non\_pkg\_begin \+ name\_len > 2

Shouldn't that be "if \(name\_len \- non\_pkg\_begin > 2"?

Hugo

**Comment by p5pRT** [comment]
### From @khwilliamson

On 7/5/19 7&#8203;:23 AM\, Hugo van der Sanden via RT wrote&#8203;:
> On Thu\, 27 Jun 2019 12&#8203;:31&#8203;:30 \-0700\, public@&#8203;khwilliamson\.com wrote&#8203;:
>> I'm trying to figure out if this is a real potential attack vector\.
>> This is malloc'd space at the top of the function\.  It's mallocd to
>> name\_len bytes\, but isn't necessarily initialized to the full amount\.
>>
>> I don't understand how this actually is generating this valgrind error\.
>> The underlying macro expands to a memcmp of name\_len bytes\, which
>> shouldn't go off the end of the mallocd space\.  But it could certainly
>> read data that didn't get initialized\.
> 
> I think lookup\_name has been moved since it was malloced\, probably by&#8203;:
>    lookup\_name \+=  STRLENs\("utf8&#8203;::"\);
> 
> \.\. which would explain valgrind's diagnosis&#8203;:
>    Address 0x51c8026 is 6 bytes inside a block of size 7
> 
> In any case\, reading \(but not leaking information about\) some bytes of malloc bookkeeping doesn't sound to me like much of an attack vector\. I think worst case is that we get unlucky with the bookkeeping bytes causing the memBEGINPs to return TRUE; I'd expect that to be only a bug though\, not a security issue\.
> 
> Briefly looking at other uses of name\_len\, this one might also be dodgy&#8203;:
> 
>      /\* If the original input began with 'In' or 'Is'\, it could be a subroutine
>       \* call to a user\-defined property instead of a Unicode property name\. \*/
>      if \(    non\_pkg\_begin \+ name\_len > 2
> 
> Shouldn't that be "if \(name\_len \- non\_pkg\_begin > 2"?

Yes it should\, thank you\.  Fixed in 8d3e0237887e7149be56d17a9448cb465edc5f76

Since no one has disagreed with Hugo's assessment that this isn't a 
security issue\, I have pushed a fix in
c6f37d61adb39a6d7764b1bcb6bb67accb22a0a0

and moved this to the public queue\, marking it pending release\.

Is a test needed\, since this only surfaces under valgrind?

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'open' to 'pending release'
